File: src/python/xerxes/acp/transport.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Self-contained stdio JSON-RPC 2.0 transport for the ACP server.

Newline-delimited JSON (one object per line) over stdin/stdout — no third-party
dependency required, so ``xerxes-acp`` always works. Requests dispatch to
:class:`AcpServer` methods; ``prompt`` runs on a worker thread so ``cancel`` and
``respond_permission`` can be processed *while a prompt streams*. Streamed agent
events are delivered as ``session/update`` notifications carrying the ACP wire
event; the prompt's final summary is returned as the JSON-RPC response.

Method aliases accept both the ``snake_case`` core names and ``slash/`` forms.
"""

from __future__ import annotations

import json
import logging
import sys
import threading
from typing import IO, Any

from .runner import AcpAgentRunner
from .server import AcpServer

logger = logging.getLogger(__name__)

# JSON-RPC 2.0 error codes.
_PARSE_ERROR = -32700
_INVALID_REQUEST = -32600
_METHOD_NOT_FOUND = -32601
_INTERNAL_ERROR = -32603


class StdioJsonRpcServer:
    """Drive an :class:`AcpServer` over newline-delimited JSON-RPC on stdio."""

    def __init__(
        self,
        server: AcpServer,
        runner: AcpAgentRunner | None = None,
        *,
        stdin: IO[str] | None = None,
        stdout: IO[str] | None = None,
    ) -> None:
        """Bind the server/runner and the stdio streams (defaults to real stdio)."""
        self._server = server
        self._runner = runner
        self._stdin = stdin or sys.stdin
        self._stdout = stdout or sys.stdout
        self._write_lock = threading.Lock()
        self._workers: set[threading.Thread] = set()
        self._running = True

    # ----- wire helpers -----------------------------------------------------

    def _send(self, obj: dict[str, Any]) -> None:
        """Serialise ``obj`` as one JSON line to stdout (thread-safe)."""
        line = json.dumps(obj, default=str)
        with self._write_lock:
            self._stdout.write(line + "\n")
            self._stdout.flush()

    def _result(self, req_id: Any, result: Any) -> None:
        self._send({"jsonrpc": "2.0", "id": req_id, "result": result})

    def _error(self, req_id: Any, code: int, message: str) -> None:
        self._send({"jsonrpc": "2.0", "id": req_id, "error": {"code": code, "message": message}})

    def _notify(self, method: str, params: dict[str, Any]) -> None:
        self._send({"jsonrpc": "2.0", "method": method, "params": params})

    # ----- main loop --------------------------------------------------------

    def serve_forever(self) -> None:
        """Read JSON-RPC lines until EOF, dispatching each request."""
        for raw in self._stdin:
            if not self._running:
                break
            line = raw.strip()
            if not line:
                continue
            self._handle_line(line)
        self._running = False
        for worker in list(self._workers):
            worker.join(timeout=5.0)

    def _handle_line(self, line: str) -> None:
        """Parse and dispatch a single JSON-RPC line; never raises."""
        try:
            msg = json.loads(line)
        except json.JSONDecodeError as exc:
            self._error(None, _PARSE_ERROR, f"parse error: {exc}")
            return
        if not isinstance(msg, dict) or msg.get("jsonrpc") != "2.0" or "method" not in msg:
            self._error(msg.get("id") if isinstance(msg, dict) else None, _INVALID_REQUEST, "invalid request")
            return
        req_id = msg.get("id")
        method = str(msg.get("method", ""))
        params = msg.get("params") or {}
        if not isinstance(params, dict):
            self._error(req_id, _INVALID_REQUEST, "params must be an object")
            return
        try:
            self._dispatch(req_id, method, params)
        except Exception as exc:  # noqa: BLE001 — one bad call must not kill the server
            logger.warning("ACP dispatch error for %s: %s", method, exc)
            if req_id is not None:
                self._error(req_id, _INTERNAL_ERROR, f"{type(exc).__name__}: {exc}")

    def _dispatch(self, req_id: Any, method: str, params: dict[str, Any]) -> None:
        """Route ``method`` to the matching server call (snake_case or slash form)."""
        srv = self._server
        m = method.replace("/", "_").lower()

        if m in ("initialize",):
            self._result(req_id, srv.initialize(params.get("client_info")))
        elif m in ("list_tools", "tools_list"):
            self._result(req_id, srv.list_tools())
        elif m in ("list_models", "models_list"):
            self._result(req_id, srv.list_models())
        elif m in ("open_session", "session_new", "session_open"):
            self._result(
                req_id,
                srv.open_session(str(params.get("cwd", "")), model=params.get("model"), title=str(params.get("title", ""))),
            )
        elif m in ("list_sessions", "session_list"):
            self._result(req_id, srv.list_sessions())
        elif m in ("set_model", "session_set_model"):
            self._result(req_id, srv.set_model(str(params.get("session_id", "")), params.get("model")))
        elif m in ("cancel", "session_cancel"):
            self._result(req_id, srv.cancel(str(params.get("session_id", ""))))
        elif m in ("close_session", "session_close"):
            self._result(req_id, srv.close_session(str(params.get("session_id", ""))))
        elif m in ("respond_permission", "permission_respond"):
            self._result(req_id, srv.respond_permission(str(params.get("permission_id", "")), bool(params.get("allow", False))))
        elif m in ("pending_permissions", "permission_pending"):
            self._result(req_id, srv.pending_permissions())
        elif m in ("respond_question", "question_respond", "respond_input", "input_respond"):
            if self._runner is None:
                self._result(req_id, {"ok": False})
            else:
                self._result(req_id, self._runner.respond_question(str(params.get("input_id", "")), str(params.get("answer", ""))))
        elif m in ("pending_questions", "question_pending", "pending_inputs"):
            self._result(req_id, self._runner.pending_questions() if self._runner else [])
        elif m in ("prompt", "session_prompt"):
            self._start_prompt(req_id, params)
        elif m in ("shutdown", "exit"):
            self._running = False
            self._result(req_id, {"ok": True})
        else:
            if req_id is not None:
                self._error(req_id, _METHOD_NOT_FOUND, f"unknown method: {method}")

    # ----- prompt (runs on a worker thread) ---------------------------------

    def _start_prompt(self, req_id: Any, params: dict[str, Any]) -> None:
        """Validate the session, then stream the prompt on a worker thread."""
        session_id = str(params.get("session_id", ""))
        text = str(params.get("text", params.get("prompt", "")))
        session = self._server.sessions.get(session_id)
        if session is None:
            self._error(req_id, _INVALID_REQUEST, f"unknown session: {session_id}")
            return
        if self._runner is None:
            self._error(req_id, _INTERNAL_ERROR, "no agent runner wired")
            return

        def emit(event: dict[str, Any]) -> None:
            self._notify("session/update", {"session_id": session_id, "request_id": req_id, "event": event})

        def work() -> None:
            try:
                summary = self._runner.run_prompt(session=session, text=text, emit=emit)
                self._result(req_id, summary)
            except Exception as exc:  # noqa: BLE001
                logger.warning("ACP prompt worker failed: %s", exc)
                self._error(req_id, _INTERNAL_ERROR, f"{type(exc).__name__}: {exc}")
            finally:
                self._workers.discard(threading.current_thread())

        worker = threading.Thread(target=work, name=f"acp-prompt-{session_id[:8]}", daemon=True)
        self._workers.add(worker)
        worker.start()


__all__ = ["StdioJsonRpcServer"]

File: xerxes/src/acp/transport.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import { errorMessage } from '../executors/toolRegistry.js'
import { AcpServer } from './server.js'
import {
  ACP_JSON_RPC_ERRORS,
  acpJsonRpcFailure,
  acpJsonRpcSuccess,
  isAcpJsonRpcId,
  isAcpRecord,
  type AcpJsonRpcId,
  type AcpWireEvent,
} from './types.js'

export type AcpStdioWriter = (line: string) => void | Promise<void>

/** Matches the daemon socket frame cap: one NDJSON frame may not exceed 16 MiB. */
export const MAX_ACP_FRAME_BYTES = 16 * 1024 * 1024

/** Incrementally accounts for bytes in the current NDJSON frame. */
export class AcpFrameByteCounter {
  private currentBytes = 0
  private scannedBytes = 0

  constructor(private readonly maxBytes = MAX_ACP_FRAME_BYTES) {}

  append(chunk: Uint8Array): void {
    for (const byte of chunk) {
      this.scannedBytes += 1
      this.currentBytes = byte === 0x0a ? 0 : this.currentBytes + 1
    }
  }

  get exceedsLimit(): boolean {
    return this.currentBytes > this.maxBytes
  }

  /** Exposed so fragmented-input tests can assert accounting remains linear. */
  get bytesScanned(): number {
    return this.scannedBytes
  }
}

interface ParsedRequest {
  readonly canReply: boolean
  readonly id: AcpJsonRpcId
  readonly method: string
  readonly params: Record<string, unknown>
}

type Sender = (frame: unknown) => Promise<void>

/**
 * Newline-delimited JSON-RPC 2.0 transport for an ACP server.
 *
 * A `prompt` is intentionally dispatched on an independent async worker so
 * `cancel`, `respond_permission`, and `respond_question` frames remain
 * processable while the prompt emits incremental `session/update` events.
 */
export class StdioJsonRpcServer {
  private readonly promptShutdown = new AbortController()
  private readonly workers = new Set<Promise<void>>()
  private running = true
  private writeQueue: Promise<void> = Promise.resolve()

  constructor(private readonly server: AcpServer) {}

  async serve(input: ReadableStream<Uint8Array>, write: AcpStdioWriter): Promise<void> {
    const reader = input.getReader()
    const decoder = new TextDecoder()
    const send = this.serializedSender(write)
    const frameBytes = new AcpFrameByteCounter()
    let buffer = ''
    try {
      while (this.running) {
        const { done, value } = await reader.read()
        if (done) {
          break
        }
        frameBytes.append(value)
        buffer += decoder.decode(value, { stream: true })
        buffer = await this.handleCompleteLines(buffer, send)
        if (this.running && frameBytes.exceedsLimit) {
          await send(acpJsonRpcFailure(
            null,
            ACP_JSON_RPC_ERRORS.invalidRequest,
            `frame exceeds maximum size of ${MAX_ACP_FRAME_BYTES} bytes`,
          ))
          this.running = false
          await reader.cancel().catch(() => undefined)
          break
        }
      }
      buffer += decoder.decode()
      if (this.running && buffer.trim()) {
        await this.handleLine(buffer.trim(), send)
      }
    } finally {
      reader.releaseLock()
      this.stopPrompts()
      await Promise.allSettled([...this.workers])
      await this.writeQueue
    }
  }

  private serializedSender(write: AcpStdioWriter): Sender {
    return frame => {
      const line = `${JSON.stringify(frame)}\n`
      const sent = this.writeQueue.then(() => write(line)).then(() => undefined)
      this.writeQueue = sent.catch(() => undefined)
      return sent
    }
  }

  private async handleCompleteLines(buffer: string, send: Sender): Promise<string> {
    let newline = buffer.indexOf('\n')
    while (newline >= 0) {
      const rawLine = buffer.slice(0, newline)
      buffer = buffer.slice(newline + 1)
      if (utf8ByteLength(rawLine) > MAX_ACP_FRAME_BYTES) {
        await send(acpJsonRpcFailure(
          null,
          ACP_JSON_RPC_ERRORS.invalidRequest,
          `frame exceeds maximum size of ${MAX_ACP_FRAME_BYTES} bytes`,
        ))
        this.running = false
        return buffer
      }
      const line = rawLine.trim()
      if (line) {
        await this.handleLine(line, send)
      }
      if (!this.running) {
        return buffer
      }
      newline = buffer.indexOf('\n')
    }
    return buffer
  }

  private async handleLine(line: string, send: Sender): Promise<void> {
    let value: unknown
    try {
      value = JSON.parse(line) as unknown
    } catch (error) {
      await send(acpJsonRpcFailure(null, ACP_JSON_RPC_ERRORS.parseError, `parse error: ${errorMessage(error)}`))
      return
    }
    if (!isAcpRecord(value) || value.jsonrpc !== '2.0' || typeof value.method !== 'string') {
      const id = isAcpRecord(value) && isAcpJsonRpcId(value.id) ? value.id : null
      await send(acpJsonRpcFailure(id, ACP_JSON_RPC_ERRORS.invalidRequest, 'invalid request'))
      return
    }

    const hasId = Object.hasOwn(value, 'id')
    if (hasId && !isAcpJsonRpcId(value.id)) {
      await send(acpJsonRpcFailure(null, ACP_JSON_RPC_ERRORS.invalidRequest, 'JSON-RPC request id must be a string, number, or null'))
      return
    }
    const id = hasId ? value.id as AcpJsonRpcId : null
    // JSON-RPC notifications are identified by an omitted id. Although null
    // ids are discouraged by the spec, they are still requests and receive a
    // null-id response for compatibility with clients that use them.
    const canReply = hasId
    if (value.params !== undefined && !isAcpRecord(value.params)) {
      if (canReply) {
        await send(acpJsonRpcFailure(id, ACP_JSON_RPC_ERRORS.invalidParams, 'params must be an object'))
      }
      return
    }
    const request: ParsedRequest = {
      id,
      canReply,
      method: value.method,
      params: value.params ?? {},
    }
    try {
      await this.dispatch(request, send)
    } catch (error) {
      if (request.canReply) {
        await send(acpJsonRpcFailure(request.id, ACP_JSON_RPC_ERRORS.internalError, `${errorMessage(error)}`))
      }
    }
  }

  private async dispatch(request: ParsedRequest, send: Sender): Promise<void> {
    const method = request.method.replaceAll('/', '_').toLowerCase()
    switch (method) {
      case 'initialize':
        await this.result(request, this.server.initialize(recordParameter(request.params.client_info)), send)
        return
      case 'list_tools':
      case 'tools_list':
        await this.result(request, this.server.listTools(), send)
        return
      case 'list_models':
      case 'models_list':
        await this.result(request, this.server.listModels(), send)
        return
      case 'open_session':
      case 'session_new':
      case 'session_open':
        {
          const model = modelParameter(request.params.model)
        await this.result(request, this.server.openSession(stringParameter(request.params.cwd), {
          ...(model !== undefined ? { model } : {}),
          ...(typeof request.params.title === 'string' ? { title: request.params.title } : {}),
        }), send)
        return
        }
      case 'list_sessions':
      case 'session_list':
        await this.result(request, this.server.listSessions(), send)
        return
      case 'set_model':
      case 'session_set_model':
        await this.result(request, this.server.setModel(stringParameter(request.params.session_id), modelParameter(request.params.model) ?? null), send)
        return
      case 'cancel':
      case 'session_cancel':
        await this.result(request, this.server.cancel(stringParameter(request.params.session_id)), send)
        return
      case 'close_session':
      case 'session_close':
        await this.result(request, this.server.closeSession(stringParameter(request.params.session_id)), send)
        return
      case 'respond_permission':
      case 'permission_respond':
        if (typeof request.params.allow !== 'boolean') {
          if (request.canReply) {
            await send(acpJsonRpcFailure(
              request.id,
              ACP_JSON_RPC_ERRORS.invalidParams,
              'respond_permission requires a boolean allow parameter',
            ))
          }
          return
        }
        await this.result(request, this.server.respondPermission(
          stringParameter(request.params.permission_id),
          request.params.allow,
        ), send)
        return
      case 'pending_permissions':
      case 'permission_pending':
        await this.result(request, this.server.pendingPermissions(), send)
        return
      case 'respond_question':
      case 'question_respond':
      case 'respond_input':
      case 'input_respond':
        await this.result(request, this.server.respondQuestion(
          stringParameter(request.params.input_id),
          stringParameter(request.params.answer),
        ), send)
        return
      case 'pending_questions':
      case 'question_pending':
      case 'pending_inputs':
        await this.result(request, this.server.pendingQuestions(), send)
        return
      case 'prompt':
      case 'session_prompt':
        await this.startPrompt(request, send)
        return
      case 'shutdown':
      case 'exit':
        this.running = false
        await this.result(request, { ok: true }, send)
        this.stopPrompts()
        return
      default:
        if (request.canReply) {
          await send(acpJsonRpcFailure(request.id, ACP_JSON_RPC_ERRORS.methodNotFound, `unknown method: ${request.method}`))
        }
    }
  }

  private async result(request: ParsedRequest, result: unknown, send: Sender): Promise<void> {
    if (!request.canReply) {
      return
    }
    await send(acpJsonRpcSuccess(request.id, result))
  }

  private async startPrompt(request: ParsedRequest, send: Sender): Promise<void> {
    const sessionId = stringParameter(request.params.session_id)
    const session = this.server.sessions.get(sessionId)
    if (!session) {
      if (request.canReply) {
        await send(acpJsonRpcFailure(request.id, ACP_JSON_RPC_ERRORS.invalidRequest, `unknown session: ${sessionId}`))
      }
      return
    }
    const text = typeof request.params.text === 'string'
      ? request.params.text
      : stringParameter(request.params.prompt)
    const worker = this.runPromptWorker(request, sessionId, text, send)
    this.workers.add(worker)
    void worker.then(
      () => this.workers.delete(worker),
      () => this.workers.delete(worker),
    )
  }

  private async runPromptWorker(request: ParsedRequest, sessionId: string, text: string, send: Sender): Promise<void> {
    let acceptsUpdates = true
    const prompt = this.server.prompt(sessionId, text, async event => {
      if (!acceptsUpdates) {
        return
      }
      await send(sessionUpdate(sessionId, request.canReply ? request.id : null, event))
    })
    const outcome = await Promise.race([
      prompt.then(
        result => ({ kind: 'result' as const, result }),
        error => ({ kind: 'error' as const, error }),
      ),
      aborted(this.promptShutdown.signal).then(async () => {
        // Let an already-settling handler publish its contractual response;
        // only detach workers that remain pending after shutdown cancellation.
        await new Promise(resolve => setTimeout(resolve, 0))
        return { kind: 'shutdown' as const }
      }),
    ])
    acceptsUpdates = false
    if (outcome.kind === 'shutdown' || !request.canReply) {
      return
    }
    if (outcome.kind === 'result') {
      await send(acpJsonRpcSuccess(request.id, outcome.result))
      return
    }
    await send(acpJsonRpcFailure(request.id, ACP_JSON_RPC_ERRORS.internalError, errorMessage(outcome.error)))
  }

  private stopPrompts(): void {
    this.server.shutdown()
    if (!this.promptShutdown.signal.aborted) {
      this.promptShutdown.abort()
    }
  }
}

/** Serve an ACP JSON-RPC endpoint from injected Bun/WHATWG stdio streams. */
export async function serveACPStdio(
  server: AcpServer,
  input: ReadableStream<Uint8Array>,
  write: AcpStdioWriter,
): Promise<void> {
  await new StdioJsonRpcServer(server).serve(input, write)
}

function aborted(signal: AbortSignal): Promise<void> {
  if (signal.aborted) {
    return Promise.resolve()
  }
  return new Promise(resolve => signal.addEventListener('abort', () => resolve(), { once: true }))
}

function utf8ByteLength(value: string): number {
  return new TextEncoder().encode(value).byteLength
}

function recordParameter(value: unknown): Record<string, unknown> | undefined {
  return isAcpRecord(value) ? value : undefined
}

function stringParameter(value: unknown): string {
  return typeof value === 'string' ? value : ''
}

function modelParameter(value: unknown): string | null | undefined {
  if (value === null || typeof value === 'string') {
    return value
  }
  return undefined
}

function sessionUpdate(sessionId: string, requestId: AcpJsonRpcId, event: AcpWireEvent): Record<string, unknown> {
  return {
    jsonrpc: '2.0',
    method: 'session/update',
    params: {
      session_id: sessionId,
      request_id: requestId,
      event,
    },
  }
}
